src/analysis/processing/qgsalgorithmfuzzifyraster.cpp

```diff
@@ -0,0 +1,561 @@
+  prepareAlgorithmFuzzificationParameters( parameters, context, feedback );
+
+  return true;
+}
+
+
+QVariantMap QgsFuzzifyRasterAlgorithmBase::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  std::unique_ptr< QgsRasterFileWriter > writer = qgis::make_unique< QgsRasterFileWriter >( outputFile );
+  writer->setOutputProviderKey( QStringLiteral( "gdal" ) );
+  writer->setOutputFormat( outputFormat );
+  std::unique_ptr<QgsRasterDataProvider > provider( writer->createOneBandRaster( mDataType, mNbCellsXProvider, mNbCellsYProvider, mExtent, mCrs ) );
+  if ( !provider )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output: %1" ).arg( outputFile ) );
+  if ( !provider->isValid() )
+    throw QgsProcessingException( QObject::tr( "Could not create raster output %1: %2" ).arg( outputFile, provider->error().message( QgsErrorMessage::Text ) ) );
+
+  provider->setNoDataValue( 1, mNoDataValue );
+  mDestinationRasterProvider = provider.get();
+  mDestinationRasterProvider->setEditable( true );
+
+  qgssize layerSize = static_cast< qgssize >( mLayerWidth ) * static_cast< qgssize >( mLayerHeight );
+  int maxWidth = QgsRasterIterator::DEFAULT_MAXIMUM_TILE_WIDTH;
+  int maxHeight = QgsRasterIterator::DEFAULT_MAXIMUM_TILE_HEIGHT;
+  int nbBlocksWidth = static_cast< int >( std::ceil( 1.0 * mLayerWidth / maxWidth ) );
+  int nbBlocksHeight = static_cast< int >( std::ceil( 1.0 * mLayerHeight / maxHeight ) );
+  int nbBlocks = nbBlocksWidth * nbBlocksHeight;
+
+  QgsRasterIterator iter( mInterface.get() );
+  iter.startRasterRead( mBand, mLayerWidth, mLayerHeight, mExtent );
+  int iterLeft = 0;
+  int iterTop = 0;
+  int iterCols = 0;
+  int iterRows = 0;
+  bool isNoData = false;
+  std::unique_ptr< QgsRasterBlock > rasterBlock;
+  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, rasterBlock, iterLeft, iterTop ) )
+  {
+    if ( feedback )
+      feedback->setProgress( 100 * ( ( iterTop / maxHeight * nbBlocksWidth ) + iterLeft / maxWidth ) / nbBlocks );
+    std::unique_ptr< QgsRasterBlock > fuzzifiedBlock = qgis::make_unique< QgsRasterBlock >( mDestinationRasterProvider->dataType( 1 ), iterCols, iterRows );
+
+    for ( int row = 0; row < iterRows; row++ )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+      for ( int column = 0; column < iterCols; column++ )
+      {
+        if ( feedback && feedback->isCanceled() )
+          break;
+        double value = rasterBlock->valueAndNoData( row, column, isNoData );
+        if ( isNoData )
+        {
+          fuzzifiedBlock->setValue( row, column, mNoDataValue );
+        }
+        else
+        {
+          double fuzzifiedValue = fuzzify( value );
```

Comment:
While it would result in duplicate code, for maximum performance you may want to consider making the fuzzify virtual method encompass this whole block of code which includes the row/column iteration too. This would avoid the 1000's of per-pixel calls to the virtual fuzzify method, and allow the compiler to potentially optimise these loops for each individual subclass.